app/src/main/java/com/infomaniak/mail/ui/main/move/MoveFragment.kt

```diff
@@ -101,79 +110,90 @@ class MoveFragment : MenuFoldersFragment() {
-                mainViewModel.moveToNewFolder(folderName, navigationArgs.threadsUids.toList(), navigationArgs.messageUid)
+                mainViewModel.moveToNewFolder(folderName, threadsUids.toList(), messageUid)
             },
             onErrorCheck = ::checkForFolderCreationErrors,
         )
     }
 
-    private fun setupFolderAdapters() {
-        moveViewModel.getFolderIdAndCustomFolders().observe(viewLifecycleOwner) { (folderId, customFolders) ->
+    private fun observeFolders() {
+        moveViewModel.getCurrentFolderAndAllFolders().observe(viewLifecycleOwner) { (allFolders, currentFolderId) ->
+            moveAdapter.setFolders(allFolders, currentFolderId)
+            setupSearchBar(allFolders, currentFolderId)
+        }
+    }
 
-            currentFolderId = folderId
+    private fun observeSearchResults() {
+        moveViewModel.filterResults.observe(viewLifecycleOwner) { (filteredFolders, currentFolderId) ->
+            moveAdapter.setFolders(filteredFolders, currentFolderId)
+        }
+    }
 
-            val defaultFoldersWithoutDraft = mainViewModel.currentDefaultFoldersLive.value!!.let { folders ->
-                folders.filterNot { it.role == FolderRole.DRAFT }
-            }
+    private fun observeFolderCreation() = with(mainViewModel) {
 
-            defaultFoldersAdapter.setFolders(defaultFoldersWithoutDraft, folderId)
-            customFoldersAdapter.setFolders(customFolders, folderId)
-            setSearchBarUi(allFolders = defaultFoldersWithoutDraft + customFolders)
+        newFolderResultTrigger.observe(viewLifecycleOwner) {
+            inputDialog.resetLoadingAndDismiss()
         }
-    }
 
-    private fun observeNewFolderCreation() = with(mainViewModel) {
-        newFolderResultTrigger.observe(viewLifecycleOwner) { inputDialog.resetLoadingAndDismiss() }
         isMovedToNewFolder.observe(viewLifecycleOwner) { isFolderCreated ->
             if (isFolderCreated) findNavController().popBackStack()
         }
     }
 
-    override fun onFolderSelected(folderId: String): Unit = with(navigationArgs) {
+    private fun onFolderSelected(folderId: String): Unit = with(navigationArgs) {
         mainViewModel.moveThreadsOrMessageTo(folderId, threadsUids.toList(), messageUid)
         findNavController().popBackStack()
     }
 
-    override fun onFolderCollapse(folderId: String, shouldCollapse: Boolean) = Unit
+    /**
+     * Asynchronously validate folder name locally
+     * @return error string, otherwise null
+     */
+    private fun checkForFolderCreationErrors(folderName: CharSequence): String? = when {
+        folderName.length > 255 -> getString(R.string.errorNewFolderNameTooLong)
+        folderName.contains(Regex(INVALID_CHARACTERS_PATTERN)) -> getString(R.string.errorNewFolderInvalidCharacter)
+        folderController.getRootFolder(folderName.toString()) != null -> context?.getString(R.string.errorNewFolderAlreadyExists)
+        else -> null
+    }
```

Comment:
This function is duplicated, can't we like, extract it as an extension (original idea not by me)

Reply:
Also, the Regex could have been instanciated a single time instead of being instanciated each time the check is made at every character, if it's not to complicated to achieve. And if this can be done using a `by lazy` even better